### PR TITLE
Implement literals

### PR DESCRIPTION
**Pull request by swazrgb**
Scripts can assign numbers and entities using literal numbers & strings:

```typescript
let num: Value = 10;
let id: Value = "c_battery";
```

Entity & number pairs can be assigned using the `value` builtin function, or as a casted object.

```typescript
let pair1 = value("c_battery", 10);
let pair2 = {
  id: "c_battery",
  num: 10
} as ItemNum;
```

Coordinates can be assigned using the `coord` builtin function, or as a casted 2-element number array.

```typescript
let coord1 = coord(10, 20);
let coord2 = [10, 20] as Coord;
```

One drawback of the new typing system is that arthimetrics aren't allowed by default, and the `Value` type has to be coerced to a number first. Fortunately this is easy to do:

```typescript
let v: Value = "c_battery";
v = +v + +v; // prefix with +
v = v as number + 1; // cast
v = 1 + v; // reverse the expression if adding a constant
```

### CODE REVIEW

**Comment by swazrgb** [comment]
@ribrdb ping

I've tried various alternatives since, mostly to work around having to write `const v: Value = +v + +v` for math, but nothing works well while still allowing literals such as `"c_battery"` or `5` to be assigned to `Value` typed variables.

The best alternative would be requiring the usage of the `value` function in all cases, so `const v: Value = value("c_battery")` or `const v: Value = value(5)`

**Comment by ribrdb** [comment]
> I've tried various alternatives since, mostly to work around having to write `const v: Value = +v + +v` for math, but nothing works well while still allowing literals such as `"c_battery"` or `5` to be assigned to `Value` typed variables: `const v: Value = 5`.

I was thinking of just adding `num: number` to the `BaseVar` interface so you could optionally use `v = v.num + v.num` instead of unary-plus. I haven't tested to see if that would work though.

> The best alternative would be requiring the usage of the `value` function in all cases, so `const v: Value = value("c_battery")` or `const v: Value = value(5)`

Actually I kind of like that

**Comment by swazrgb** [comment]
> I was thinking of just adding num: number to the BaseVar interface so you could optionally use v = v.num + v.num instead of unary-plus. I haven't tested to see if that would work though.

That should work. It feels a bit odd to me to provide a fake property for this. Internally the `.num` accessor would just be a no-op, existing purely to get the typescript to compile.

> Actually I kind of like that

I don't mind it for variable assignments, but it's awkward for function calls. For example having to write `entitiesInRange(5, value("v_mineable"))` would be unfortunate.

This seems doable:

```typescript
let v: Value; // Declare variable/params as Value
v = 5; // numbers are directly assignable
v = v + 5; // Values & numbers can be used in math as-is
v = value("metalore"); // string assingments must be wrapped in `value` 
v = value("metalore", 5); // NumPairs need the value function as well of course

v = "metalore"; // Not allowed! Must use value("metalore");

entitiesInRange(5, "v_mineable"); // literal strings can be supported as function arguments
```

**Comment by swazrgb** [comment]
That was pretty straightforward. Could you please take another look, @ribrdb ?

File: compile.ts [hunk 1]
```diff
@@ -603,9 +603,88 @@ class Compiler {
...
+      }
+      return dest;
     } else if (ts.isParenthesizedExpression(e)) {
       return this.compileExpr(e.expression, dest);
+    } else if (ts.isAsExpression(e)) {
+      return this.compileExpr(e.expression, dest);
+    } else if (ts.isPrefixUnaryExpression(e)) {
+      if(e.operator == ts.SyntaxKind.PlusToken) {
+        return this.compileExpr(e.operand, dest);
+      } else {
```
Comment (ribrdb) [inline compile.ts:685]: Should we support unary `-`? (e.g compile it to a subtraction op)

File: compile.ts [hunk 2]
```diff
@@ -603,9 +603,88 @@ class Compiler {
...
+          value,
+          this.ref(dest, VariableOperations.Write)
+        );
+      } else {
+        return new Variable(value);
+      }
+      return dest;
+    } else if (ts.isArrayLiteralExpression(e)) {
+      let arr: number[] = [];
+      for(const property of e.elements) {
```
Comment (ribrdb) [inline compile.ts:619]: Why is this needed when there's the `coord` function?
If we do need it we should probably check the array length. I could certainly see people trying to use general arrays and getting confused.

File: compile.ts [hunk 3]
```diff
@@ -717,6 +796,65 @@ class Compiler {
...
+  builtins: Record<string, (e: ts.CallExpression) => LiteralValue> = {
+    value: (e) => {
+      if (e.arguments.length < 1 || e.arguments.length > 2) {
+        this.#error(`Unsupported argument length: ${e.arguments.length}`, e);
+      }
+
+      const a = this.parseBuiltinArg(e.arguments[0]);
+
+      if (typeof a !== 'string') {
+        this.#error(`Unsupported argument type for argument 2: (${ts.SyntaxKind[e.arguments[0].kind]})`, e.arguments[0]);
```
Comment (ribrdb) [inline compile.ts:818]:
```suggestion
        this.#error(`Unsupported argument type for argument 1: (${ts.SyntaxKind[e.arguments[0].kind]})`, e.arguments[0]);
```

**Comment by ribrdb** [comment]
This looks great, thanks! Just a couple comments

File: compile.ts (hunk 1 above)
Comment (swazrgb) [reply compile.ts:685]: I'd like to, and I have a branch that does this, but I'd like to combine it with literal evaluating math ops before emitting them.

So we could let a user write

```
const v: Value = -1
```

And we could just compile this to `-1` instead of a `sub` operation which "wastes" an instruction (which matters in locked mode, which is the only way to play on the recent servers).

So yes, I'd like to, and I'd like to PR this + my inlining behavior soon. Needs a bunch of cleanup, but a sneak preview: https://github.com/swazrgb/desynced-tools/blob/inlining/compile.ts#L1157-L1178

Might be a good idea to discuss this in a new issue, but do you have any thoughts for the right place to perform such optimizations? During the compilation phase? In a pass afterwards? In the assembler?

File: compile.ts (hunk 2 above)
Comment (swazrgb) [reply compile.ts:619]: I implemented it because I saw the typescript definition `type Coord = [number, number]` which implies a user should be allowed to write this when specifying a coordinate. But I'd be happy to just get rid of it, the `coord` function is better ergonomics IMO.

Comment (swazrgb) [reply compile.ts:619]: Maybe change the `Coord` type to `{x: number, y: number}`, and allow this to be written literally? Just like `{id: "metalore", num: 1}` is also allowed to be written literally?

Or disallow writing both literally, and require usage of the `coord` and `value` functions? Then the type could change to:

```
class Coord {
  x: number;
  y: number
}
```

File: compile.ts (hunk 1 above)
Comment (ribrdb) [reply compile.ts:685]: I think that'd be best in a pass after compilation and before assembly.

File: compile.ts (hunk 2 above)
Comment (ribrdb) [reply compile.ts:619]: Ah, I was only intending that to be used for arguments to functions which take a coordinate so you could enter a literal. You're right, an `{x:, y:}` object would probably be more clear.

**Review by ribrdb — APPROVED** [review]
